layout: Make the compositor rather than layout determine the position of each iframe.

pcwalton opened the pull request:
The old code that attempted to do this during layout wasn't able to work
for multiple reasons: it couldn't know where the iframe was going to be
on the page (because of nested iframes), and at the time it was building
the display list for a fragment it couldn't know where that fragment was
going to be in page coordinates.

This patch rewrites that code so that only the size of an iframe is
determined during layout, and the position is determined by the
compositor. Layout layerizes iframes and marks the iframe layers with
the appropriate subpage ID so that the compositor can place them
correctly.

Closes #7377.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7423)

<!-- Reviewable:end -->

pcwalton committed:
layout: Make the compositor rather than layout determine the position of
each iframe.

The old code that attempted to do this during layout wasn't able to work
for multiple reasons: it couldn't know where the iframe was going to be
on the page (because of nested iframes), and at the time it was building
the display list for a fragment it couldn't know where that fragment was
going to be in page coordinates.

This patch rewrites that code so that both the sizes and positions of
iframes are determined by the compositor. Layout layerizes all iframes
and marks the iframe layers with the appropriate pipeline and subpage
IDs so that the compositor can place them correctly. This approach is
similar in spirit to Gecko's `RefLayer` infrastructure. The logic that
determines when it is time to take the screenshot for reftests has been
significantly revamped to deal with this change in delegation of
responsibility.

Additionally, this code removes the infrastructure that sends layout
data back to the layout task to be destroyed, since it is now all
thread-safe and can be destroyed on the script task.

The failing tests now fail because of a pre-existing bug related to
intrinsic heights and borders on inline replaced elements. They happened
to pass before because we never rendered the iframes at all, which meant
they never had a chance to draw the red border the tests expect to not
render!

Closes #7377.

pcwalton commented:
r? @glennw

glennw commented on components/compositing/compositor.rs:
```diff
@@ -644,9 +650,11 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+        if let Some(ref frame_size) = frame_size {
+            let frame_size = frame_size.to_untyped();
+            root_layer.bounds.borrow_mut().size = Size2D::from_untyped(&frame_size);
+            /*root_layer.bounds.borrow_mut().origin =
```
Can be removed

glennw commented on components/compositing/constellation.rs:
```diff
@@ -382,11 +382,17 @@ impl<LTF: LayoutTaskFactory, STF: ScriptTaskFactory> Constellation<LTF, STF> {
+                self.handle_frame_size_msg(pipeline_id, subpage_id, &Size2D::from_untyped(&size));
+            }
+            ConstellationMsg::SetSubpagePosition(pipeline_id, subpage_id, position) => {
+                debug!("constellation got frame rect message");
```
comment needs updating

glennw commented on components/layout/display_list_builder.rs:
```diff
@@ -1262,10 +1255,12 @@ impl FragmentDisplayListBuilding for Fragment {
         let layer = match layer {
             StackingContextLayer::Existing(existing_layer) => Some(existing_layer),
             StackingContextLayer::IfCanvas(layer_id) => {
```
Perhaps rename ::IfCanvas() if it's now used for things other than canvas?

glennw commented on components/layout/display_list_builder.rs:
```diff
@@ -1814,21 +1818,28 @@ impl InlineFlowDisplayListBuilding for InlineFlow {
-            } else {
-                !self.fragments.fragments[0].style().get_effects().filter.is_empty()
+        //
+        // FIXME(pcwalton): This is terribly, terribly bogus! What is even going on here?
```
Can we add an issue for this - it looks very odd.

pcwalton replied:
Filed #7424.

glennw commented:
@pcwalton Just a few minor nits, looks good otherwise.

pcwalton commented:
@bors-servo: r=glennw

bors-servo commented:
:pushpin: Commit 791ec95 has been approved by `glennw`

<!-- @bors-servo r=glennw 791ec9581ad77168ba4db17536a770bd76f42efe -->

bors-servo commented:
:hourglass: Testing commit 791ec95 with merge 0356328...

bors-servo commented:
:broken_heart: Test failed - [mac2](http://build.servo.org/builders/mac2/builds/3155)

pcwalton commented:
@bors-servo: r=glennw

bors-servo commented:
:pushpin: Commit a6737e4 has been approved by `glennw`

<!-- @bors-servo r=glennw a6737e439d1c949990a2123e307927696715172c -->

bors-servo commented:
:hourglass: Testing commit a6737e4 with merge 8e5f862...

bors-servo commented:
:broken_heart: Test failed - [mac3](http://build.servo.org/builders/mac3/builds/1109)

pcwalton commented:
Added some more expected test failures. I'm pretty sure the failures are pre-existing for the reasons stated in the commit message. Out of an abundance of caution I'd like to rerequest review. r? @glennw

glennw commented:
@pcwalton Looking at some of those wpt/css failures, it's not immediately clear to me why they would start failing - could you provide some more info?

pcwalton commented:
I fixed the failing test cases. Was a pretty simple fix to add `LAYERS_NEEDED_FOR_POSITIONED_DESCENDANTS`. Carrying forward review.

pcwalton commented:
@bors-servo: r=glennw

bors-servo commented:
:pushpin: Commit 37a69f5 has been approved by `glennw`

<!-- @bors-servo r=glennw 37a69f569b9f0b9b0026c14161bc51bde314285f -->

bors-servo commented:
:hourglass: Testing commit 37a69f5 with merge ec5f565...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/29)

jdm commented:
```

command timed out: 1200 seconds without output running ['./mach', 'test-ref', '--kind', 'cpu'], attempting to kill
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7313) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton commented:
I believe the races are fixed now.

r? @glennw

glennw commented:
Reviewed 14 of 16 files at r1.
Review status: 0 of 23 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/compositing/constellation.rs, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7423#-Jxm24ce40HS_1sINEuj-r1-392)** ([raw file](https://github.com/servo/servo/blob/d4d87ed32cf04f908c2470ca2270483da869db4e/components/compositing/constellation.rs#L392)):</sup>
comment needs updating

---

<sup>**[components/layout/display_list_builder.rs, line 1257 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7423#-Jxm24ce40HS_1sINEuo-r1-1257)** ([raw file](https://github.com/servo/servo/blob/d4d87ed32cf04f908c2470ca2270483da869db4e/components/layout/display_list_builder.rs#L1257)):</sup>
Perhaps rename ::IfCanvas() if it's now used for things other than canvas?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)
<!-- Sent from Reviewable.io -->

glennw commented:
Reviewed 2 of 16 files at r1, 18 of 21 files at r2.
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
Why is ReapLayoutData not needed anymore?

---

<sup>**[components/script/dom/window.rs, line 803 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwznFqyZqFKJyDqAmg)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/window.rs#L803)):</sup>
If this comment can be removed, the GC should be too?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)
<!-- Sent from Reviewable.io -->

glennw commented:
@pcwalton This generally looks OK to me (it seems to match what we discussed at the work week).

I'm not sure that ReapLayoutData can be removed though? There are a few nits apart from that question, and test-tidy needs to be fixed.

It's really unfortunate that the subpage abstraction is now leaking into the compositor, but I don't have any suggestion for a better way to fix it right now (other than fixing #5235 and #5241) :(

pcwalton commented:
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
The `.unwrap()` line was causing trouble here because the layout task could have been shut down. It didn't seem like it was actually doing anything anymore, since all `handle_reap_layout_data` does is call `layout_data_wrapper.remove_compositor_layers` (which is now gone) so I just removed it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)
<!-- Sent from Reviewable.io -->

glennw commented:
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
There is the line "let layout_data_wrapper: LayoutDataWrapper = transmute(layout_data);" in handle_reap_layout_data() that does a transmute to a different type. I *think* this causes memory to be freed correctly, which may otherwise leak without reap layout data. Perhaps @jdm could comment on that?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)
<!-- Sent from Reviewable.io -->

pcwalton commented:
@glennw New version of the patch restores the layout data reaping. Embarrassing that I forgot about that :)

glennw commented:
@pcwalton Looks good to me once travis is happy (which I assume is a test-tidy failure but the travis results won't load for me for some reason).

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7643) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton commented:
@bors-servo: r=glennw

bors-servo commented:
:pushpin: Commit 002e120 has been approved by `glennw`

<!-- @bors-servo r=glennw 002e120aeb4040566becce7495f7db4c0efafb3a -->

bors-servo commented:
:hourglass: Testing commit 002e120 with merge da904a1...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/284)

jdm commented:
```

/css-flexbox-1_dev/html/flexbox-basic-iframe-horiz-001.htm
----------------------------------------------------------
TIMEOUT expected FAIL [Parent]
/css-flexbox-1_dev/html/flexbox-basic-iframe-vert-001.htm
---------------------------------------------------------
TIMEOUT expected FAIL [Parent]
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7771) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton commented:
OK, I believe I've fixed those tests. The failures came from a deadlock in the way the compositor blocked on the constellation, which could happen simultaneously with the constellation waiting on the compositor. Fixing this by introducing a new message pair (`PrepareForSubpageLayerCreation` and `CreateLayerForSubpage`) exposed another race in detecting whether screenshots were safe to take, which I fixed by modifying the constellation; see the comment above `if let Some(size) = pipeline.size` in `constellation.rs`.

I also effectively redid #7771 with an alternate approach in order to fix the rebase conflict.

r? @glennw

glennw commented:
Reviewed 3 of 21 files at r2, 12 of 12 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)
<!-- Sent from Reviewable.io -->

glennw commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit c72d0c2 has been approved by `glennw`

<!-- @bors-servo r=glennw c72d0c2ed04f6c5f9fa22a3f576e1df5878f941d -->

bors-servo commented:
:hourglass: Testing commit c72d0c2 with merge 29c061e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/340)

mbrubeck commented:
@bors-servo retry

#7787

bors-servo commented:
:hourglass: Testing commit c72d0c2 with merge a0cb657...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/3229), [gonk](http://build.servo.org/builders/gonk/builds/2389), [linux-dev](http://build.servo.org/builders/linux-dev/builds/306), [linux-rel](http://build.servo.org/builders/linux-rel/builds/320), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/317), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/315), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/342)
